VoodooInput/VoodooInputWellspringSimulator/VoodooInputWellspringSimulator.hpp

```diff
@@ -0,0 +1,172 @@
+//
+//  AppleUSBMultitouchDriver.hpp
+//  VoodooInput
+//
+//  Created by Avery Black on 12/31/22.
+//  Copyright © 2022 Kishor Prins. All rights reserved.
+//
+
+#ifndef AppleUSBMultitouchDriver_hpp
+#define AppleUSBMultitouchDriver_hpp
+
+#include <IOKit/hid/IOHIDDevice.h>
+#include <IOKit/hid/IOHIDEventService.h>
+#include <IOKit/IOLib.h>
+#include <IOKit/IOKitKeys.h>
+#include <IOKit/IOWorkLoop.h>
+#include <IOKit/IOCommandGate.h>
+
+#include "../VoodooInput.hpp"
+#include "../VoodooInputMultitouch/VoodooInputTransducer.h"
+#include "../VoodooInputMultitouch/VoodooInputEvent.h"
+#include "../VoodooInputMultitouch/MultitouchHelpers.h"
+
+#define MT1_MAX_X 8920
+#define MT1_MAX_Y 6600
+
+#define MT1_LITTLE_ENDIAN 1
+
+/* State bits reference: linux/drivers/hid/hid-magicmouse.c#L58-L63 */
+#define MT2_TOUCH_STATE_BIT_TRANSITION (0x1)
+#define MT2_TOUCH_STATE_BIT_NEAR (0x1 << 1)
+#define MT2_TOUCH_STATE_BIT_CONTACT (0x1 << 2)
+
+class VoodooInputWellspringSimulator;
+class VoodooInputWellspringUserClient;
+
+class AppleUSBMultitouchHIDEventDriver : public IOHIDEventService {
+    OSDeclareDefaultStructorsWithDispatch(AppleUSBMultitouchHIDEventDriver);
+    friend class VoodooInputWellspringSimulator;
+public:
+    virtual IOReturn setSystemProperties(OSDictionary *) override;
+};
+
+// This report does not come from the hardware, but instead comes from within AppleUSBMultitouch in 10.12+
+// This gets sent to userspace on any button presses
+struct __attribute__((__packed__)) MTRelativePointerReport {
+    UInt8 ReportID;
+    UInt8 Unknown1; // Always set to one
+    UInt16 Unknown2;
+    UInt32 Buttons;
+    UInt32 Unknown3[4]; // A dx/dy probably exists in here
+    AbsoluteTime Timestamp;
+};
+
+struct __attribute__((__packed__)) WELLSPRING3_FINGER {
+    UInt8 Id;
+    UInt8 State;
+    UInt8 Finger;
+    UInt8 Unknown; // 0 -> 1 -> 0xFF
+    SInt16 X;
+    SInt16 Y;
+    // Velocity can be calculated in MultitouchSupport.plugin for us
+    // as long as "AlwaysNeedVelocityCalculated" is set.
+    SInt16 XVelocity;
+    SInt16 YVelocity;
+    UInt16 ToolMajor;
+    UInt16 ToolMinor;
+    UInt16 Orientation;
+    UInt16 Size;
+    // ContactDensity can be calculated in MultitouchSupport.plugin
+    /*UInt16 ContactDensity;
+    UInt16 Unused[2];
+    UInt16 Pressure; // Not a thing on Wellspring3 */
```

Comment:
These would need to be uncommented to add pressure support, and may need to change the product ID/other device properties. As long as we give the correct sizes in the header though, older versions should still parse this fine.